### ISSUE TITLE
Upload file for attachment

### ISSUE THREAD

**Issue by zapadi** [zapadi/redmine-net-api#356]
### Discussed in https://github.com/zapadi/redmine-net-api/discussions/355

<div type='discussions-op-text'>

<sup>Originally posted by **joewak** August 22, 2024</sup>
I'm trying this:
```
            byte[] documentData = System.IO.File.ReadAllBytes(file_path);
            Upload attachment = manager.UploadFile(documentData);
```

When I do so, I get an exception: 
```
           WebException: The remote server returned an error: (422) Unprocessable Entity.
           RedmineException: Invalid or missing attribute parameters: Attachment extension  is not allowed
```

Poking around in this function in InternalRedmineApiWebClient.cs:
```
        public ApiResponseMessage Upload(string address, byte[] data, RequestOptions requestOptions = null)
        {
            var content = new ByteArrayApiRequestMessageContent(data);
            return HandleRequest(address, HttpVerbs.**UPLOAD**, requestOptions, content);
        }

```
The Redmine REST API documentation says a POST should be used to upload a file.  Should it be POST instead of UPLOAD?

</div>

**Comment by zapadi** [comment]
Published version `4.6.7`

**Comment by joewak** [comment]
I tested with version 4.6.7.  Now I get:

`"The remote server returned an error: (406) Not Acceptable."`

**Comment by joewak** [comment]
I was able to upload and attach a file.

I made this function which went outside of your library to upload the file using HttpClient to do the POST operation.  Using this function I was able to get the 201 response code and get the **token** from the response.
```
public static bool UploadFile(string file_path, out string response_token)
        {
            bool retval = false;
            response_token = "";

            try
            {
                byte[] documentData = System.IO.File.ReadAllBytes(file_path);
                string filename = System.IO.Path.GetFileName(file_path);

                using (var content = new ByteArrayContent(documentData))
                {
                    content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                    content.Headers.Add("X-Redmine-API-Key", apiKey);

                    var client = new HttpClient();
                    var task1 = Task.Run(() => client.PostAsync("https://ptcobservations.wabtec.com/uploads.xml?filename=" + filename, content));
                    task1.Wait();
                    var response = task1.Result;
                    var task2 = Task.Run(() => response.Content.ReadAsStringAsync());
                    task2.Wait();
                    var response_contents = task2.Result;

                    XmlDocument xml = new XmlDocument();
                    xml.LoadXml(response_contents);

                    response_token = xml.ChildNodes[1].ChildNodes[0].ChildNodes[0].Value;

                    if (response_token != "")
                    {
                        retval = true;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            return retval;
        }
```

Then I was able to use the **token** I received with your library to attach the file to my issue:

```
public static bool AttachFile(string token, string filename, string ticket_id)
        {
            bool retval = false;

            try
            {
                RedmineManager manager = new RedmineManager(host, apiKey);

                Upload attachment = new Upload();

                ////set attachment properties
                attachment.FileName = filename;
                attachment.Description = "Remote download package for " + ticket_id;
                attachment.ContentType = "application/octet-stream";
                attachment.Token = token;

                ////create a list of attachments to be added to issue
                IList<Upload> attachments = new List<Upload>();
                attachments.Add(attachment);

                var parameters = new NameValueCollection { };
                Issue issue = manager.GetObject<Issue>(ticket_id, parameters);
                issue.Uploads = attachments;

                if (issue != null)
                {
                    manager.UpdateObject(issue.Id.ToString(), issue);
                    retval = true;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            return retval;
        }
```

**Comment by zapadi** [comment]
Fixed in 4.6.9 version

**Comment by joewak** [comment]
Sorry, I still don't think it works.  Back to 422 error.
